pkg/plugins/policies/meshretry/plugin/v1alpha1/plugin.go

```diff
@@ -26,55 +30,94 @@ func (p plugin) MatchedPolicies(dataplane *core_mesh.DataplaneResource, resource
+func (p plugin) Apply(rs *core_xds.ResourceSet, ctx xds_context.Context, proxy *core_xds.Proxy) error {
 	policies, ok := proxy.Policies.Dynamic[api.MeshRetryType]
 	if !ok {
 		return nil
 	}
 
 	listeners := policies_xds.GatherListeners(rs)
+	routes := policies_xds.GatherRoutes(rs)
 
 	if err := applyToOutbounds(policies.ToRules, listeners.Outbound, proxy.Dataplane, proxy.Routing); err != nil {
 		return err
 	}
 
+	if err := applyToGateway(ctx, policies.ToRules, routes.Gateway, listeners.Gateway, proxy); err != nil {
+		return err
+	}
+
 	return nil
 }
 
-func applyToOutbounds(rules core_xds.ToRules, outboundListeners map[mesh_proto.OutboundInterface]*envoy_listener.Listener, dataplane *core_mesh.DataplaneResource, routing core_xds.Routing) error {
+func applyToOutbounds(
+	rules core_xds.ToRules,
+	outboundListeners map[mesh_proto.OutboundInterface]*envoy_listener.Listener,
+	dataplane *core_mesh.DataplaneResource,
+	routing core_xds.Routing,
+) error {
 	for _, outbound := range dataplane.Spec.Networking.GetOutbound() {
-		serviceName := outbound.GetTagsIncludingLegacy()[mesh_proto.ServiceTag]
 		oface := dataplane.Spec.Networking.ToOutboundInterface(outbound)
+		serviceName := outbound.GetTagsIncludingLegacy()[mesh_proto.ServiceTag]
+
+		configurer := plugin_xds.Configurer{
+			Retry:    core_xds.ComputeConf[api.Conf](rules.Rules, core_xds.MeshService(serviceName)),
+			Protocol: policies_xds.InferProtocol(routing, serviceName),
+		}
+
 		listener, ok := outboundListeners[oface]
 		if !ok {
 			continue
 		}
 
-		protocol := policies_xds.InferProtocol(routing, serviceName)
-
-		if err := configure(rules.Rules, core_xds.MeshService(serviceName), protocol, listener); err != nil {
+		if err := configurer.ConfigureListener(listener); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-func configure(rules core_xds.Rules, subset core_xds.Subset, protocol core_mesh.Protocol, listener *envoy_listener.Listener) error {
-	var conf api.Conf
-	if computed := rules.Compute(subset); computed != nil {
-		conf = computed.Conf.(api.Conf)
-	} else {
+func applyToGateway(
+	ctx xds_context.Context,
+	rules core_xds.ToRules,
+	gatewayRoutes map[string]*envoy_route.RouteConfiguration,
+	gatewayListeners map[core_xds.InboundListener]*envoy_listener.Listener,
+	proxy *core_xds.Proxy,
+) error {
+	if !proxy.Dataplane.Spec.IsBuiltinGateway() {
 		return nil
 	}
 
-	configurer := plugin_xds.Configurer{
-		Retry:    &conf,
-		Protocol: protocol,
+	gatewayListerInfos, err := gateway_plugin.GatewayListenerInfoFromProxy(context.TODO(), ctx.Mesh, proxy, ctx.ControlPlane.Zone)
+	if err != nil {
+		return err
 	}
 
-	for _, filterChain := range listener.FilterChains {
-		if err := configurer.Configure(filterChain); err != nil {
+	for _, listenerInfo := range gatewayListerInfos {
+		configurer := plugin_xds.Configurer{
+			Retry:    core_xds.ComputeConf[api.Conf](rules.Rules, core_xds.MeshSubset()),
```

Comment:
Should we add a comment here/note in an issue about this assumption?

Reply:
https://github.com/kumahq/kuma/issues/6790